tapir/shifts/forms.py

```diff
@@ -192,3 +192,82 @@ class Meta:
             "start_date": DateInput(),
             "end_date": DateInput(),
         }
+
+    confirm_cancelled_attendances = BooleanField(
+        label=_(
+            "I have read the warning about the cancelled attendances and confirm that the exemption should be created"
+        ),
+        required=False,
+        widget=HiddenInput,
+    )
+    confirm_cancelled_abcd_attendances = BooleanField(
+        label=_(
+            "I have read the warning about the cancelled ABCD attendances and confirm that the exemption should be "
+            "created "
+        ),
+        required=False,
+        widget=HiddenInput,
+    )
+
+    def validate_unique(self):
+        super().validate_unique()
+        if (
+            "confirm_cancelled_attendances" in self._errors
+            or "confirm_cancelled_abcd_attendances" in self._errors
+        ):
+            return
+
+        if (
+            "confirm_cancelled_attendances" in self.cleaned_data
+            and not self.cleaned_data["confirm_cancelled_attendances"]
+        ):
+            covered_attendances = ShiftExemption.get_attendances_covered_by_exemption(
```

Comment:
`self.user.shift_attendances.overlapping_with((start_date, end_date))` should do it here. Unfortunately, `DurationModelMixin.overlapping_with` does not yet take a tuple as a date range. This was already an issue when implementing `DurationModelMixin.active_temporal`, see the hack there. We should introduce a little type check there (ugly, I know, but meh) to check if it's a tuple and if so, treat it as a date range.

Than we can also do away with that ugly static method in ShiftExemption.

Reply:
I'm not sure that works, wouldn't we need ShiftAttendance to inherit from DurationModelMixin?
Maybe we could make another version of DurationModelMixin that takes datetimes instead of dates, then have Shift inherit from that, then do something like `self.user.shift_attendances.filter(slot__shit__in=Shift.objects.overlaps_with(exemption))`

Reply:
Hm, yeah, you're right, that's a bit unlucky. I'm always a bit uncomfortable reimplementing range overlap logic, it feels like one of those edgecasey things...

Maybe you could test what sort of query Django builds from the line you posted above? If it's JOINed in the database, it's probably fine. If it ends up being two queries with a list of all shifts in the exemption being passed, I would expect it to be too slow. And since Shift and ShiftExemption are both a DurationModelMixin, it should just work, no?